### DISCUSSION TITLE
conditioning SerialWrite on change in value

### DISCUSSION THREAD

**Discussion by KathaSchmack** [bonsai-rx/bonsai#1739] [Q&A] [answered]
Hi,

we have a Bpod protocol that uses Bonsai for visual stimulation and response collection. We have a value that indicates the position of a rotary encoder. We would like Bonsai to send a certain bytes to Bpod whenever a rotary encoder changes into a certain zone, e.g. cross into center zone -> send 1 byte once, cross out of center zone -> send 2 bytes once, cross into left zone -> send 3 bytes once etc. We have this workflow that works on some computers but not on our rig computers mainly because it seems that Bpod is receiving lots and lots of bytes all the time, and our rig computers seem to be unable to deal with this. Any suggestions how to change the code so that only one byte is sent when the RotaryPosition enters the centerCoordinates?
[visualStimulation.zip](https://github.com/bonsai-rx/bonsai/files/14981219/visualStimulation.zip)
<img width="608" alt="image" src="https://github.com/bonsai-rx/bonsai/assets/95446942/cffbd6f7-3987-46e3-87e4-43f9558f504a">

Many thanks in advance!

Katharina

**Comment by bruno-f-cruz** [comment]
Hey,
Not sure exactly what you are trying to do, but the fact that all this logic runs inside a `SelectMany` makes me nervous. This operator essentialy  creates a new "workflow" each time it receives an input. If you dont close it, you will end up running N workflows in parallel without any need.
Can you describe a bit better what kind of behavior are you trying to achieve?

**Comment by KathaSchmack** [comment]
Thank you so much @bruno-f-cruz , the fact that it makes you nervous is a good indicator that this might be the problem:) 

We want to tell Bpod whenever the RotaryPosition changes in or from defined ranges:

RotaryPosition starts being inside the range [-0.1,0.1] -> send 1 byte
RotaryPosition stops being inside the range [-0.1,0.1] -> send 2 bytes 
RotaryPosition starts being inside the range [-0.6 -0.4] -> send 3 bytes
RotaryPosition stops being inside the range  [-0.6 -0.4] -> send 4 bytes
RotaryPosition starts being inside the range [0.4 0.6] -> send 5 bytes
RotaryPosition stops being inside the range [0.4 0.6] -> send 6 bytes

I hope that makes it a little clearer. I did try running it in a simple group rather than a SelectMany but that did not solve the problem.

Many thanks, Katharina

**Reply by bruno-f-cruz** [reply, marked as answer]
I think I understand now, thanks!
Would this work? I didnt change the boolean logic as I am not fully sure if the conditions overlap or not, I will leave that up to you. But basically, on value that is received from the rotary encoder, check the conditions of all branches and propagate the necessary integers that pass the condition. These integers and coverted to byte[] and sent to bpod.
[visualStimulation_refactored.zip](https://github.com/bonsai-rx/bonsai/files/14982977/visualStimulation_refactored.zip)

**Comment by KathaSchmack** [comment]
This works, thank you so much, I really appreciate this!
